### PR TITLE
Testing AM integrity

### PR DESCRIPTION
**Pull request by JackGRWD**
Attempting to test the validity of AM. Specifically, we wanted to see whether or not SF and AM were running in parallel on the port. Currently, AM thread iterates once, SF thread iterates once and then we stay on AM without ever going back to SF.

### CODE REVIEW

File: AttitudeManager/Inc/AM_States.hpp [hunk 1]
```diff
@@ -50,58 +64,82 @@ class SensorFusionMode: public AttitudeState {
...
+		return _pid_output;
+	}
 private:
-	ControlLoopMode() {}
-	ControlLoopMode(const ControlLoopMode& other);
-	ControlLoopMode& operator = (const ControlLoopMode& other);
+	ControlLoopMode() {
+	}
+	ControlLoopMode(const ControlLoopMode &other);
+	ControlLoopMode& operator =(const ControlLoopMode &other);
```
Comment (Gfcello) [inline AttitudeManager/Inc/AM_States.hpp:82]: (nit) missed a space after the =, but no functional difference

File: AttitudeManager/Inc/AM_States.hpp [hunk 2]
```diff
@@ -50,58 +64,82 @@ class SensorFusionMode: public AttitudeState {
...
+		return _channel_out;
+	}
 private:
-	OutputMixingMode() {}
-	OutputMixingMode(const OutputMixingMode& other);
-	OutputMixingMode& operator = (const OutputMixingMode& other);
+	OutputMixingMode() {
+	}
+	OutputMixingMode(const OutputMixingMode &other);
+	OutputMixingMode& operator =(const OutputMixingMode &other);
```
Comment (Gfcello) [inline AttitudeManager/Inc/AM_States.hpp:107]: (nit) same space thing

File: AttitudeManager/Inc/AM_States.hpp [hunk 3]
```diff
@@ -50,58 +64,82 @@ class SensorFusionMode: public AttitudeState {
...
+	}
 	static AttitudeState& getInstance();
 private:
-	FatalFailureMode() {}
-	FatalFailureMode(const FatalFailureMode& other);
-	FatalFailureMode& operator =(const FatalFailureMode& other);
+	FatalFailureMode() {
+	}
+	FatalFailureMode(const FatalFailureMode &other);
+	FatalFailureMode& operator =(const FatalFailureMode &other);
```
Comment (Gfcello) [inline AttitudeManager/Inc/AM_States.hpp:125]: (nit) space

File: AttitudeManager/Inc/AM_States.hpp [hunk 4]
```diff
@@ -50,58 +64,82 @@ class SensorFusionMode: public AttitudeState {
...
+	}
 	static AttitudeState& getInstance();
 private:
-	DisarmMode() {}
-	DisarmMode(const DisarmMode& other);
-	DisarmMode& operator = (const DisarmMode& other);
+	DisarmMode() {
+	}
+	DisarmMode(const DisarmMode &other);
+	DisarmMode& operator =(const DisarmMode &other);
```
Comment (Gfcello) [inline AttitudeManager/Inc/AM_States.hpp:142]: (nit) space

File: AttitudeManager/Src/AM.cpp [hunk 5]
```diff
@@ -8,29 +8,30 @@
 
 #include "../Inc/AM_Interface.h"
 
-AttitudeManager::AttitudeManager(LOS_Link *los_link, LOS_Actuators *los_output) {
-    currentState = &FetchInstructionsMode::getInstance();
-    status = AttMan::COMPLETED_CYCLE;
-    link = los_link;
-    output = los_output;
+AttitudeManager::AttitudeManager(LOS_Link *los_link,
+		LOS_Actuators *los_output) {
```
Comment (Gfcello) [inline AttitudeManager/Src/AM.cpp:12]: (nit) tab this so the LOS_Actuators is directly below LOS_Link

File: AttitudeManager/Src/AM.cpp [hunk 6]
```diff
@@ -8,29 +8,30 @@
...
+	currentState = &FetchInstructionsMode::getInstance();
+	status = AttMan::COMPLETED_CYCLE;
```
Comment (Gfcello) [inline AttitudeManager/Src/AM.cpp:14]: (nit) I dislike having both attMgr and AttMan, probably not worth changing here.

File: AttitudeManager/Src/AM_States.cpp [hunk 7]
```diff
@@ -69,63 +68,75 @@ AttitudeState& FetchInstructionsMode::getInstance() {
 }
 
 bool FetchInstructionsMode::receiveTeleopInstructions(AttitudeManager *attMgr) {
-	bool is_dc{true};
+	bool is_dc { true };
 
 	if (is_dc) {
 		return false;
 	}
```
Comment (Gfcello) [inline AttitudeManager/Src/AM_States.cpp:75]: Can you explain what this does?

File: SensorFusion/Src/SensorFusion.cpp [hunk 8]
```diff
@@ -217,66 +234,75 @@ SFError_t SF_GetAttitude(SFAttitudeOutput_t *Output, IMUData_t *imudata) {
...
+			atan2f(qDiff1 * qDiff2 + qDiff3 * qDiff4,
+					0.5f - qDiff2 * qDiff2 - qDiff3 * qDiff3)) * SF_FREQ;
+	imu_PitchRate =
+	RAD_TO_DEG(asinf(-2.0f * (qDiff2 * qDiff4 - qDiff1 * qDiff3))) * SF_FREQ;
+	imu_YawRate =
+	RAD_TO_DEG(
+			atan2f(qDiff2 * qDiff3 + qDiff1 * qDiff4,
+					0.5f - qDiff3 * qDiff3 - qDiff4 * qDiff4)) * SF_FREQ;
+
+	//Transfer Fused IMU data into SF Output struct
```
Comment (Gfcello) [inline SensorFusion/Src/SensorFusion.cpp:260]: (nit) tabbing looks weird in github

File: SensorFusion/Src/SensorFusion.cpp [hunk 9]
```diff
@@ -296,268 +322,231 @@ SFError_t SF_GetPosition(
...
+	//Confidence in the physical relationship prediction performed using f
+	float q[DIM * DIM] = { 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0,
+			0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0 };
+
+	const int16_t NUM_MEASUREMENTS = 7;
+
+	//Maps sensor data from z to x
+	float h[NUM_MEASUREMENTS * DIM] = { 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
+			0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
+			0, 0, 0, 0, 0, 1 };
```
Comment (Gfcello) [inline SensorFusion/Src/SensorFusion.cpp:361]:
```suggestion
	float h[NUM_MEASUREMENTS * DIM] = { 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
	                                                                   1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 
	                                                                   0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1 };
```

Comment (Gfcello) [reply SensorFusion/Src/SensorFusion.cpp:361]: (nit)

**Review by Gfcello — COMMENTED** [review]
From reading through this it's mostly just some formatting cleanup?

File: SensorFusion/Src/SF_Interface.cpp [hunk 10]
```diff
@@ -2,10 +2,12 @@
 
 #include "../Inc/SensorFusion.hpp"
 
+#include "../../Drivers/Sensors/Inc/imu.hpp"
```
Comment (Aidan-B) [inline SensorFusion/Src/SF_Interface.cpp:5]: Why are we including this?

File: SensorFusion/Src/SensorFusion.cpp [hunk 11]
```diff
@@ -590,55 +580,54 @@ SFError_t SF_GenerateNewResult()
...
-IMU_Data_t SF_GetRawIMU()
-{
-    IMUData_t imuData;
-    imuObj->GetResult(imuData);
-
-    IMU_Data_t imuOutput;
+IMU_Data_t SF_GetRawIMU() {
+	IMUData_t imuData;
+	//imuObj->GetResult(imuData);
+	IMU& imuObj = BMX160::getInstance();
```
Comment (Aidan-B) [inline SensorFusion/Src/SensorFusion.cpp:624]: I've noticed that the IMU has been declared a few times in local scope. Is there a reason it is no longer a global object? Is the allocation time to create/destroy these worth the ease of reading?
I can't remember if  the getInstance just returns a pointer to the same existing object, in which case this is probaby not an issue

File: SensorFusion/Src/SensorFusion.cpp [hunk 12]
```diff
@@ -296,268 +322,231 @@ SFError_t SF_GetPosition(
...
-                    1, 0, 0, 0, 0, 0,
-                    0, 1, 0, 0, 0, 0,
-                    0, 0, 1, 0, 0, 0,
-                    0, 0, 0, 1, 0, 0,
-                    0, 0, 0, 0, 1, 0,
-                    0, 0, 0, 0, 0, 1
-            };
+	//The identity matrix
+	float I[DIM * DIM] = { 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
+			0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1 };
```
Comment (Aidan-B) [inline SensorFusion/Src/SensorFusion.cpp:499]: Formatting same

File: SensorFusion/Src/SensorFusion.cpp [hunk 13]
```diff
@@ -296,268 +322,231 @@ SFError_t SF_GetPosition(
...
+	//Relationship between distance and acceleration
+	float ddt = pow(dt, 2) / 2;
+
+	//Maps u to x, incorporating acceleration into the estimate
+	float b[DIM * U_DIM] = { ddt, 0, 0, dt, 0, 0, 0, ddt, 0, 0, dt, 0, 0, 0,
+			ddt, 0, 0, dt };
+
+	//Confidence in the physical relationship prediction performed using f
+	float q[DIM * DIM] = { 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0,
+			0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0, 0, 0, 0, 0, 0, 0, 5.0 };
```
Comment (Aidan-B) [inline SensorFusion/Src/SensorFusion.cpp:354]: Same formatting comment as above

File: SensorFusion/Src/SensorFusion.cpp [hunk 14]
```diff
@@ -296,268 +322,231 @@ SFError_t SF_GetPosition(
...
-        h[6*DIM+5] = 0;
-    }
+	/* Matrix Definitions */
+
+	//Number of variables being estimated (dimension of the state vector x).
+	const int16_t DIM = 6;
+
+	//Maps x to itself, applying any physical relationships between variables.
+	float f[DIM * DIM] = { 1, dt, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, dt, 0,
+			0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, dt, 0, 0, 0, 0, 0, 1 };
```
Comment (Aidan-B) [inline SensorFusion/Src/SensorFusion.cpp:332]: I think this was easier to read before?

**Review by Aidan-B — COMMENTED** [review]
So this looks like a formatting PR, not really a testing PR. If you want to track the troubles you are having with thread switching, I would recommend you open a github issue. PRs should probably be reserved for development we want to merge into the project, and problems or bugs tracked in issues or in Asana.

As for why the thread isn't switching, I would start by making sure that bad thread is properly yielding to the scheduler. If you have done that I would recommend you describe all the symptoms you have encountered, and the steps you have taken in an attempt to resolve them. Hopefully we can get to a useful place once we know that.

**Comment by Aidan-B** [comment]
I just ran the code on a Nucleo here, I had build errors so tried commenting out the `AM_interfaceExecute()` and `SensorFusionInterfaceExecute()` functions. It seems like the threads were both running fine. I will bring your attention to the [vTaskDelayUntil()](https://www.freertos.org/vtaskdelayuntil.html) method. Notice that it will delay for the time increment such that the thread runs at a fixed frequency. Based on my quick investigation, it seems that xTimeIncrement is measured in ticks, and our project has configured 1 tick = 1 millisecond.

My suspicion is the attitude manager thread was not yielding to the scheduler because it was trying to delay until a time that had already passed (i.e. it takes longer for `AM_interfaceExecute()` to run than `xTimeIncrement` in ms). Increasing the time increment for attitudeManager may resolve the issue, or possibly both tasks. Hopefully this gives you a starting place. Good luck

**Pull request closed without merging**